README.md
![image](./Greek_Phalanx.jpg)

Phalanx is a chess playing program
Copyright (c) 1997, 1998, 1999, 2000, 2014 Dusan Dobes


LICENSE AND WARRANTY

- Phalanx is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2, or (at your option)
  any later version.
- Phalanx is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
- You should have received a copy of the GNU General Public License
  along with Phalanx; see the file COPYING.  If not, write to
  the Free Software Foundation, 59 Temple Place - Suite 330, Boston,
  MA  02111-1307, USA.


WHERE TO GET PHALANX

- http://sourceforge.net/projects/phalanx/
- ftp://sunsite.unc.edu/pub/Linux/games/strategy/   (Only sources)

  Please let me know if there are others, i will update this list.
  Feel free to put Phalanx on your ftp site.


COMPILING, PORTABILITY

Compiling is simple, at least under Linux. Just type `make'.  You might want
to edit the optimization flags in makefile to produce faster binary.  Phalanx
is developed under GNU C, GNU Debugger, and GNU Make.  If your system does
not support GNU exensions (e.g. long options), remove '-DGNUFUN' from DEFINES
in makefile.  If you have incompatible 'make', try this:
$ cat *.c > allphalanx.c; cc -O allphalanx.c -o phalanx


INTERFACE, COMMAND LINE OPTIONS

Phalanx is xboard compatible.  Running with xboard: 'xboard -fcp phalanx'.
Note that permanent brain (pondering) is off by default.  Newer versions of
xboard set it on with the 'hard' command.  If this does not work, try
'xboard -fcp "phalanx -p+"' or (for <4.0.0 versions of xboard) change your
initString (see Xboard documentation for details).  It's better to stop
permanent brain in both programs, when playing Phalanx against another
program on a machine with one CPU.
It's also possible to run phalanx without xboard. Do "phalanx -h" to get
a list of command line options. One important command of phalanx's ASCII
interface is "help".


MORE ABOUT INTERFACE

I'm trying to write an interface that fits following three requirements:
- Xboard compatibility.  For best results, get the latest version of Xboard.
- shell-like interface that allows running commands in a batch. It's very
  useful for testing. Example: look into the file test.fin. It's a set
  of chess problems and solutions. You can simply send this file to
  Phalanx's stdin:
  $ phalanx -c+ -o- -b- -f10 < test.fin | tee result
  ( Where: -c+ .... use cpu time
  -o- .... don't use polling input
  -b- .... no opening book
  -f10 ... fixed time 10 seconds per move )
  Watch how it works.
- Acceptable ASCII interface.


COMMAND LINE OPTIONS

Usage:
phalanx [options] [<moves> <minutes> [<increment in seconds>]]
phalanx [options] [<seconds per move>]
phalanx bench
phalanx bcreate [book creation options]

Options:
-t <transposition table size in kilobytes>
Transposition table really needs to fit in the RAM memory, it should
never be swapped to disk. Maximum size is your total RAM minus space
needed for OS (8-20 MB) minus space for other (possible) applications.
-f <fixed search time in seconds>
-x <+/->  xboard mode on/off                  default: on
"Xboard mode off" is designed for the ascii interface. In -s+ mode it
shows also the move that is currently searched.
-p <+/->  permanent brain on/off              default: off
Phalanx ponders on a predicted move with the permanent brain on. It is
sometimes called pondering. Xboard sets it on with the 'hard' command
regardless of this option, just use xboard menu to set it on/off.
-s <+/->  show thinking on/off                default: off
Phalanx shows depth, evaluation, time used in centiseconds, nodes
searched, and principal variation (best line) when searching. Xboard
overrides this option with 'post' and 'nopost' commands, just use
xboard menu instead of this.
-c <+/->  cpu time                            default: off
This one defaults to the "wall clock" real time. It's better to use
the CPU time when running test suites.
-o <+/->  polling input                       default: on
Use -o- for running test suites. Phalanx reads the positions from its
standard input and polling input makes it stop thinking after almost
zero seconds on each position.
-b <+/->  opening book                        default: on
-l <+/->  learning on/off                     default: off
-r <resign value in centipawns>               default: 0 (no resigning)
-e <easy level 0...99>                        default: 0 (best play)
1 is the hardest and 99 is the easiest easy level. Phalanx tries to
emulate human-like blunders, the higher the number the more blunders it
plays. It also adds more randomness with the easy levels, repeating
games should be impossible. Easy levels set hashtable size to zero,
pondering and learning to off. Since version XXIII, Phalanx uses
the time normally, it does not respond immediatelly. NPS is lowered
to 100-300, unless overriden by the -n agrument. Root moves randomizing
is used here as well, between 10 to 60 centipawns, unless overriden
by the -z argument.
-z <random evaluation in centipawns>          default: 0 (best play)
Randomize play. Add pseudo-random values to root moves evaluations.
The numeric value limits the random range, the interval of random
evaluations to be added is [-N/2 ... N/2].
-z <random evaluation in centipawns>:<N>
Like above, but only randomize first N moves. This is to avoid repeating
lines in opening, while keeping almost the same playing strength. For
example, -z 20:10 will randomize first 10 moves in the game by
20 centipawns.
-n <nodes per second>                         default: 0 (no limit)
Limits the speed to weaken the engine and to use less resources: The
speed of the machine does not matter here, it uses usleep() during
the search, so with low NPS it does not raise the machine load.
-v
Print version and exit.
-P <primary book directory>
-S <secondary book directory>
-L <learning file directory>
-g <log file name>

Book creation options:
-b <buffer size in cells>
Stage 1. One cell is 10 bytes. Twice the space is needed for sorting
the buffer, use no more than 1/3 of your total RAM.
-p <maxply>
Stage 1. Stop parsing the game at given ply (halfmove). Default value is
70, it's 35 full moves.
-c <max length of comment>
Stage 1. PGN files have comments in {}. Phalanx skips the comments when
parsing. Big PGN files also have lots of errors, sometimes '{' does not
have matching '}'. So, if the comment goes over the maximal length, we
suspect the text is no more the comment and continue parsing. Default
value is 4096 bytes.
-w <winning move value>
Stage 1. We are adding the [position,move] to the database. The side to
move won the game. The appropriate database entry gets <winning move
value> bonus. Default is 5 points. Use values in 1-10 range.
-d <drawing move value>
Stage 1. Same as the previous except that the game was drawn. Default
value is 2 points.
-l <losing move value>
Stage 1. Same as the previous except that the side to move lost the
game. Default value is 1 point.
-u <unknown result move value>
Stage 1. Same as the previous except that there is no result of the game
in the PGN file, sad and common case. Default value is 1 point.
-v <min value of move to add>
Stage 2. Moves that have less total points than this value are not
added into the sbook.phalanx. Bigger numbers create smaller books.
The smallest reasonable number is 10. Default is 15.
-g <min value percentage of best move to add others>
Stage 2. Use values in 50-100 range. The 2nd, 3rd, and other moves are
stripped if their value is less than G% of the value of the best move.
The G is decreased in positions where the 1st move value has high enough
value to give more variability in frequently played opening, e.g. we
want a bit more than just sicilian, maybe 1. ... e5 is also good.
50 gives lots of variability and maybe dubious moves, 100 gives almost
no variability and only the highest valued moves. Default is 80.

Note that the book creation does not work with DOS end-of-lines in the PGN
input, just use dos2unix filter if needed.

Examples: phalanx -c+ -s+ -o - -x- -f 60 -t4000
xboard -fcp "phalanx -r800"
xboard -fcp 'phalanx -e 99'


ASCII INTERFACE COMMANDS

about            about Phalanx, show version, copyright info, and settings
bd               display position (same as 'd')
bk               show book moves (xboard) - shows ECO code/name and all moves
from both primary and secondary opening books. The last line
is the text string that is used to identify the position in
the pbook.phalanx. Primary book moves are shown along with
their probabilities. There is no such probability info for
the secondary book moves, as all secondary book moves have
equal probability.
book             enable/disable opening book
both             machine plays both
depth            set search depth in plies (xboard). Search will be stopped
at given depth, no timing info is used.
fen              display position in FEN
force            user plays both (xboard)
go               switch sides, start computing (xboard)
history          show game moves in full notation
level N M I      set level to N moves in M minutes, increment I seconds.
Phalanx needs its time info have updated from xboard via
the 'time' command to make this level work well.
level N          set level to fixed time N seconds per move
new              new game (xboard)
post             show thinking (xboard)
remove           take back last move, two plies (xboard)
nopost           do not show thinking (xboard)
quit             quit (same as 'exit' or end of file character)
score            show static evaluation
time <N>         remaining time is N/100 s (xboard)
undo             undo last ply (xboard; same as 'u')
<FEN position>   set test position, start search, show result
#                comment


OPENING BOOK

From version VI, there are two book files - primary (pbook.phalanx), and
secondary (sbook.phalanx).  A position is first searched in pbook.phalanx.
Only if it's not found there, sbook.phalanx is searched.  You can specify
book directories via command line (-P, -S) or use environment variables
PHALANXPBOOKDIR and PHALANXSBOOKDIR.  Otherwise Phalanx tries to find its
book files in current directory (./book.phalanx, ./sbook.phalanx) and finally
in compiled-in directory (/usr/local/lib/phalanx).  You can change the
compiled-in directory in makefile.
- pbook.phalanx is 'hand'-written, text book.  One line per position, sorted.
  This time, it's bigger than really needed, because it was the only book
  file till version V.  The size will be smaller and the line format might
  change to EPD+SAN in future.
- sbook.phalanx is binary book, generated from large PGN files.  Six bytes
  per move (4 hash key, 2 move).  You can generate your own sbook.phalanx
  with 'phalanx bcreate <options>', like this:
  $ ./phalanx bcreate < manyGMgames.pgn
  Book creation has two stages. First stage reads and parses PGN from
  standard input and creates the 'rbook.phalanx' file. If rbook.phalanx
  already exists, the first stage is skipped. Second stage filters positions
  and moves from the rbook.phalanx to sbook.phalanx. First stage needs about
  30 times more time than the second one. Look at 'book creation options' for
  more details on creating your sbook.phalanx.


ECO DATABASE

From version XX, Phalanx has ECO database.  Command 'bk' can also print ECO
code and opening/variation name.  This works with xboard 4.0 via
"Help"/"Book" menu item.  If you want to activate this feature, you must type
the 'bk' command (or click the "Help"/"Book" menu item in xboard) once per
session in the initial position.  Then, the ECO index is created (that might
take few seconds on a slow machine).  The ECO index is based on positions
rather than on moves so transpositions between openings can take place.  The
eco.phalanx file is taken from Gnuchess distribution (originally eco.pgn, but
this is not real pgn).


INSIDE THE MACHINE

Phalanx uses (traditional) 10x12 board implementation.  There are three
often used board implementations: "8x8" (GNU Chess), "bitboard" (Crafty),
and "10x12" (Nimzo, Phalanx).  In short, "10x12" is easy to implement and
the code and basic data structures are small ( == fast on PC).  The engine
uses many well known techniques: PVS (principal variation search),
transposition/killer table, static-eval cache, history killers, SEE (static
exchange evaluator), null move pruning, forward pruning, internal iterative
deepening, chess-specific extensions.


AUTHOR

Dusan Dobes


phalanx/HISTORY
This is history of Phalanx.  Dates are in the form YYYYMMDD.

I         First version to play a legitimate game, it finally knows all rules.
19970304  I spent a lot of time on tuning transposition table code and
          extensions. The program scores 78% at large Reinfeld's test set, 10 s
          per move, on 486+/150.

II        Two killer heuristics added: well known history killers (modified)
199704??  and `reaction' killers (the killer is connected to the previous enemy
          move). The tree search is faster, especially at `quiet' positions.
          Hashtable presearch. Modified hashentry replacement strategy. Better
          evaluation function, it now knows more about passed pawns, the
          endgame play is much stronger. Fixed bug at 7-th row pawn push
          extension routine. Faster output - using sprintf() to prepare the
          output and then only one printf. New command line options: -t, -T to
          set the hashtable size.

III       Permanent brain (pondering).  King safety evaluation.  Hung pieces
19970529  static evaluation.  Null move pruning.  Forward pruning.  Better time
          heuristics - forced moves are played quickly.  New cmd-line options:
          -f, -x, -p, -s.

IV        New option -c to determine whether to use cpu or real time.  The
19970721  internal timing resolution changed to 1/100 of a second.  Xboard
          compatible editing position.  Removed en-passant capture from
          quiescence search.  Changes in static-eval function.  Piece list
          implemented - speedup is between 0 and 20%; more in endgame.  Pinned
          pieces static evaluation.  Better evaluation of weak pawns.  Trapped
          bishop and knight evaluation.  Bugfix at extending check evasions.
          Bugfix: development bonus was computed but not added to total
          evaluation.  Output change: '!' means turn, '!!' means value out of
          window.

V         Mostly interface changes, no big improvement in playing strength.
19970803  Xboard compatibility fixes: 'post' is no more switch, new command
          'remove'.  Fixed command 'level': 'level 0 5 0' (five minutes per
          game) should work, also increment (ICS) levels ('level 0 2 12').
          Increment can be given at command line.  New option -o (polling
          input) - when running tests, phalanx needs -o-.  New feature of
          autotesting: all incorrect results are written into file
          'notfound.fin'.  Small output changes for better compatibility with
          xboard 3.6.2.  Eliminated some bad turns in evaluation.  Better king
          safety evaluation - counting safe checks available.

VI        Fixed small efficiency bug in evaluate.c.  Improved time heuristics
19970915  for increment (ICS) levels.  Optimized hashing - about 5% overall
          speedup.  Fixed bug in ptime().  User is now allowed to continue play
          even if the position is drawn by 3-rep., material, or 50 moves rule.
          Fixed bug - Draw requests are now always ignored, not always accepted
          :-).  Xboard-compatible 'Illegal move' message.  Binary opening book,
          created from PGN by bcreate.  This distribution is bigger than any
          previous because it contains a small example of binary book
          ('sbook.phalanx', 174kB, 29634 moves).  SAN input accepted.  Static
          evaluation now better understands middlegame positions without
          castling.  Added pre-computed tables into static evaluation -> about
          5% speed improvement.  'Show thinking' mode now shows also a list of
          book moves if there's a book entry.  I have played a match with
          previous version (64 games, 2 minutes, increment 12 seconds, on
          a 486+/150, via xboard, no pondering), new version wins 38.5-25.5.

VII       Courtesy of Pavel Janik ml., new command 'fen'.  Bugfix in command
19971022  line parser, 'phalanx 0 2 12' now works.  Better time controls: added
          hard time limit to avoid being flagged.  Pondering bugfix: Phalanx
          was pondering O-O, but opponent played O-O-O; the move check was
          incorrect and Phalanx thought that O-O was played; this resulted in
          'Illegal move' message few moves later.  More pondering safety: the
          move to be pondered is now checked for legality, before this fix it
          was simply taken from PV (if present), but PV is not always 100%
          correct.  More PV safety: search cannot be interrupted during the
          first iteration, the abort is delayed until the first iteration is
          finished.  New options: -r <resign value>, -b <+/-> to set opening
          book on/off.  New commands hard/easy to set pondering on/off.
          Optimized do_move(): 3% speedup in test positions.  Search
          optimization: null move is not played if the value in hash table
          entry is too low.  Output bugfix: output is now readable even at long
          time searches.  SEE (Static Exchange Evaluator) and more pruning in
          quiescence search.  RoboFICS compatible commands 'white' and 'black'.
          Routine for evaluating pawn endgames.  New extension trick that helps
          in some horizon-effect type positions, it's based on measuring
          difference between current score and null move result; if current
          score is much better, the side to move must be under threat and the
          line is extended, similar trick is also in gnuchess.

VIII      Fixed bug in search(): eliminated useless re-searches -> about 3%
19971103  speed improvement.  Noise level is now in centiseconds, not in nodes.
          Improved horizon-effect extensions.  endgame.c: new endgame
          knowledge, mostly for endgames with minor pieces and pawns: knight
          and bishop mating, knight+knight (draw), minor vs. minor (draw),
          minor vs. pawn (draw, but still needs some work, sometimes a pawn

phalanx/phalanx.h
#ifndef PHALANX_INCLUDED
#define PHALANX_INCLUDED

#define ENGNAME "Phalanx"
#define VERSION "XXV"

#ifdef GNUFUN
# include <getopt.h>
#endif

#ifdef _WIN32
# include <windows.h>
# include <io.h>
# include <errno.h>
# include <process.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <ctype.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>

/*
 * COMPATIBILITY
 */

#ifndef CLOCKS_PER_SEC
# define CLOCKS_PER_SEC CLK_TCK
#endif

#ifdef _WIN32
# ifndef __cplusplus
# define inline __inline
# endif
  typedef __int64 int64;
#else
  typedef long long int64;
#endif

/*
 * TYPES
 */

typedef unsigned char tsquare;

/*** Moves are generated into and executed from struct tmove ***/
typedef struct
{
  short
	value;		/* heuristic value - the order key - before search */
			/* value returned - after search */
  short
	dch;		/* depth change -100...100 */
  unsigned char
	from,		/* square */
	to,		/* destination square */
	special;	/* Castling or en passant capture, see moving.c */
  unsigned char
	in1,		/* moving piece */
	in2,		/* captured piece or 0 */
	in2a;		/* moving piece or promoted piece */
} tmove;          /* 12 bytes */

typedef struct
{
  int psnl;		/* positional evaluation */
  int devi;		/* deviation to be used in lazy eval, 0 = true eval */
  int check;		/* side to move is in check */
} tsearchnode;

typedef struct
{
  tmove
	m;
  unsigned
	hashboard;
  unsigned
	rule50,		/* number of half-moves since last irreversible */
	castling;	/* 4 castling flags: */
			/*   white short, w. long, b. s., b. l. */
  int
	mtrl,		/* material count of side on move */
	xmtrl;		/* material count of opposite side */
} tgamenode;      /* 12+16 bytes */

typedef struct
{
  unsigned
	hashboard;
  short
	value,
	depth: 14;
  unsigned
	result: 2;
  unsigned short
	move,
	age;
} thashentry;     /* 12 bytes */

/* Positional features of one side */
typedef struct
{
  short hung;
  signed char khung, kshield;
  unsigned char p, n, b, r, q;   /* number of given pieces */
  unsigned char kp;              /* king position */
  unsigned char r7r;             /* # of rooks on 7-th row */
  signed char prune;             /* is it ok to prune search in this pos ? */
  signed char kstorm, qstorm;    /* diffside castling: king attack q/k side */
  unsigned char devel;           /* development needed?, 0 ... no */
  unsigned char castling;
  unsigned char bishopcolor,     /* square color of friendly bishops, or-red */
               xbishopcolor;     /* square color of enemy bishops */
  signed char worsebm;           /* mobility of the worse bishop */
  unsigned char lpf, rpf;        /* leftmost/rightmost pawn file */
} tknow;

typedef enum
	{ timecontrol, averagetime, fixedtime, fixeddepth }
tlevel;

typedef struct
{
  int	centiseconds;
  unsigned
	moves,            /* time levels */
	increment;
  int	depth;            /* fixed depth level */
  int
	noise;
  int resign;
  int
	bench;
  int
	machine_color,
	post,
	xboard,
	book,
	learn,
	cpu,           /* use cpu time */
	ponder,
	analyze,
	polling;       /* polling input */
  tlevel level;
  int nps;
  int easy;
  int random;
  int rndlim;
  FILE	* log;
} tflag;

typedef struct
{ unsigned char prev, next; }
tlist;

typedef struct
{ FILE *
	f;
  unsigned
	firstkey,
	lastkey;
  int
	filesize;
} tsbook;

typedef struct
{ FILE * f; int filesize; }
tpbook;

typedef struct
{ unsigned taxi, max, diag, min; }
tdist;



/*
 * DEFINITIONS
 */

#ifndef VERSION
# define VERSION "unknown"
#endif

/* use SAN for output */
#undef SANOUT

/* primary (text, small, user) book */
#ifndef PBOOK_FILE
#define PBOOK_FILE "pbook.phalanx"
#endif
#ifndef PBOOK_DIR
#define PBOOK_DIR "/usr/local/lib/phalanx"
#endif

/* secondary (binary, large, generated from pgn) book */
#ifndef SBOOK_FILE
#define SBOOK_FILE "sbook.phalanx"
#endif
#ifndef SBOOK_DIR
#define SBOOK_DIR "/usr/local/lib/phalanx"
#endif

#ifndef ECO_FILE
#define ECO_FILE "eco.phalanx"
#endif
#ifndef ECO_DIR
#define ECO_DIR "/usr/local/lib/phalanx"
#endif

#ifndef LEARN_FILE
#define LEARN_FILE "learn.phalanx"
#endif
#ifndef LEARN_DIR
#define LEARN_DIR "/var/local/lib/phalanx"
#endif

#define ENV_PBOOK "PHALANXPBOOKDIR"
#define ENV_SBOOK "PHALANXSBOOKDIR"
#define ENV_LEARN "PHALANXLEARNDIR"
#define ENV_ECO "PHALANXECODIR"

#define max(x,y) (((x)>(y))?(x):(y))
#define min(x,y) (((x)<(y))?(x):(y))

#define NOVALUE 32123
#define CHECKMATE 30000
#define MAXPLY 64
#define MAXCOUNTER 1024 /* 512 moves */

#define HASH_COLOR ((unsigned)0xFDB97531)
#define no_cut     1
#define alpha_cut  2
#define beta_cut   3

/* Castling special move codes and flags */

#define SHORT_CASTLING 1
#define LONG_CASTLING 2
#define NULL_MOVE 20

#define WSHORT 1 /* if set, white short castling is impossible */
#define WLONG  2
#define BSHORT 4
#define BLONG  8

/* Colors / sides */

#define WHITE 1
#define BLACK 2

#define empty(X) (((X)&3)==0)
#define white(X) (((X)&3)==WHITE)
#define black(X) (((X)&3)==BLACK)
#define color(X) ((X)&3)
#define enemy(X) ((X)^3)

/* Values */

#define P_VALUE  100
#define N_VALUE  350
#define B_VALUE  350
#define R_VALUE  550
#define Q_VALUE 1050

/* Pieces */

#define piece(X) ((X)&0x7C)

#define PAWN   0x10
#define KNIGHT 0x20
#define BISHOP 0x30
#define ROOK   0x40
#define QUEEN  0x50
#define KING   0x60

#define WP (WHITE|PAWN)
#define WN (WHITE|KNIGHT)
#define WB (WHITE|BISHOP)
#define WR (WHITE|ROOK)
#define WQ (WHITE|QUEEN)
#define WK (WHITE|KING)

#define BP (BLACK|PAWN)
#define BN (BLACK|KNIGHT)
#define BB (BLACK|BISHOP)
#define BR (BLACK|ROOK)
#define BQ (BLACK|QUEEN)
#define BK (BLACK|KING)

/*** Bit masks for power tables ***/

#define WPM 0x0001
#define WNM 0x0002
#define WBM 0x0004
#define WRM 0x0008
#define WQM 0x0010
#define WWW 0x0020

#define BPM 0x0100
#define BNM 0x0200
#define BBM 0x0400
#define BRM 0x0800
#define BQM 0x1000
#define BBB 0x2000

/* Files */

#define FILE_A 1
#define FILE_B 2
#define FILE_C 3
#define FILE_D 4
#define FILE_E 5
#define FILE_F 6
#define FILE_G 7
#define FILE_H 8

/* Squares */

#define A1 21
#define B1 22
#define C1 23
#define D1 24
#define E1 25
#define F1 26
#define G1 27
#define H1 28

#define A2 31
#define B2 32
#define C2 33
#define D2 34
#define E2 35
#define F2 36
#define G2 37
#define H2 38

#define A3 41
#define B3 42
#define C3 43
#define D3 44
#define E3 45
#define F3 46
#define G3 47
#define H3 48

#define A4 51
#define B4 52
#define C4 53
#define D4 54
#define E4 55
#define F4 56
#define G4 57
#define H4 58

#define A5 61
#define B5 62
#define C5 63
#define D5 64
#define E5 65
#define F5 66
#define G5 67
#define H5 68

#define A6 71
#define B6 72
#define C6 73
#define D6 74
#define E6 75
#define F6 76
#define G6 77
#define H6 78

#define A7 81
#define B7 82
#define C7 83
#define D7 84
#define E7 85
#define F7 86
#define G7 87
#define H7 88

#define A8 91
#define B8 92
#define C8 93
#define D8 94
#define E8 95
#define F8 96
#define G8 97
#define H8 98
#define H9 99 /* a bumper: we are out of board */


/*
 * VARIABLES
 */

extern tflag Flag;                    /* Misc. settings, see typedef of tflag */

extern tsquare B[120];         /* The chessboard */
extern tlist L[120];                  /* List of pieces */
#define WKP (L[1].next)               /* White king position */
#define BKP (L[2].next)               /* Black king position */

extern signed char * Th;              /* horizontal axe transf., (H8<->H1) */
extern signed char * Tv;              /* vertical axe tr., (H8<->A8) */

extern const char initialpos[];

extern tdist dist[120*120];
extern tknow Wknow, Bknow;            /* Things i know about actual position */
extern int Color;                /* c. to move: 1 ... white, 2 ... black */
extern int LastIter;                /* result of last iteration */
extern int Depth;                   /* depth to be searched, in % of ply */
extern int Ply;
extern int FollowPV;
extern int Totmat;                  /* total mtrl sum on the board */
extern int Abort, NoAbort;          /* abort the search when SIG_INT */
extern long AllDepth;                 /* Statistics for average Depth */
extern int64 AllNPS;
extern long T1, T2;                 /* time started & planned on a move */

extern int DrawScore;
#define DRAW ( Ply%2 ? -DrawScore : DrawScore )

extern tsearchnode S[MAXPLY];       /* the current node is S[Ply] */
extern tgamenode G[MAXCOUNTER];
extern int Counter;                 /* game counter, points to G[] */

extern int64 Nodes;           /* statistics */
extern int Scoring;
extern int EasyMove;

extern long Otim;                     /* remaining opponent's time */

/* move generation tables */
extern int N_moves[8], RB_dirs[8];
#define K_moves RB_dirs

extern int Values[7];            /* piece values */
extern unsigned P[120];

extern tmove PV[MAXPLY][MAXPLY];   /* principal variantion */
extern tmove Pondermove;

/* Hashing codes */
extern unsigned H[12][64];
extern thashentry *HT;             /* Hash table itself */
extern int * HS;                 /* square trans-table */
extern int HP[];                 /* piece trans-table */
extern unsigned SizeHT;            /* Size of the hash table */
extern unsigned Age;

/* Opening book */
extern tpbook Pbook;               /* primary book flags */
extern tpbook Learn;
extern tsbook Sbook;               /* secondary book flags */
extern FILE * Eco;
extern int Bookout;

/* params that cannot be pushed thru SIGALM handler, */
/* must be pushed thru global variables */
extern volatile int A_n, A_i, A_d;
extern volatile tmove * A_m;



/*
 * FUNCTIONS
 */

extern void myfwrite( void*, int, FILE* );
extern void myfread( void*, int, FILE* );

extern int bcreate(int,char**);

extern void new_game(void);

/* input/output */
extern int printfl(const char *, ...);
extern void printm(tmove,char*);
extern void printmSAN( tmove*, int, int, char* );
extern void printPV(int,int,char*);
extern void infoline(int,char*);
extern void verboseline(void);
extern void printboard(char*);
extern int setfen(const char*);
extern void shell(void);
extern int command(void);
extern tmove * sandex(char*,tmove*,int);

/* signal handlers */
extern void interrupt(int);

extern int attacktest(int,int);
#define checktest(side) attacktest( (side)==WHITE ? WKP : BKP, enemy(side))

extern void generate_legal_moves(tmove*,int*,int);
extern void generate_legal_captures(tmove*,int*,int);
extern void generate_legal_checks(tmove*,int*);
extern int see( tsquare *, int, int );

extern int search(tmove*,int,int,int);
extern tmove root_search(void);

extern int repetition(int);
extern int material_draw(void);
extern int evaluate(int,int);
extern void blunder(tmove*,int*);

extern int score_position(void);

/* endgame knowledge */
extern int pawns(void);
extern int e_nb(int);           /* endgame: knight and bishop */
extern int e_mp(void);            /*          minor piece and pawn */
extern int e_rpr(void);           /*          R+P vs. R */

extern void do_move( tmove * );
extern void undo_move( tmove * );

extern unsigned hashboard(void);
extern thashentry * seekHT(void);
extern void writeHT(int,int,int);

/* levels */
extern void l_level(char *);
extern void l_startsearch(void);
extern int l_iterate(void);
extern long ptime(void);
extern long LastTurn;
extern int Turns;

/* book */
extern int bookmove(tmove*,int);
extern unsigned smove(tmove*);

/* learning feature */
extern int rlearn(void);
extern void wlearn(int,int);

/* killer move utils */
extern void init_killers(void);
extern void write_killer( int, int );
extern void add_killer( tmove *, int, thashentry * );
extern void slash_killers( tmove *, int );

#endif

